src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs

```diff
@@ -170,8 +161,146 @@ public Task<RestoreResult> ExecuteAsync()
+            {
+                foreach (var cycle in g.AnalyzeResult.Cycles)
+                {
+                    logger.LogError(Strings.Log_CycleDetected + $" {Environment.NewLine}  {cycle.GetPath()}.");
+                    return false;
+                }
+
+                foreach (var versionConflict in g.AnalyzeResult.VersionConflicts)
+                {
+                    logger.LogError(
+                        string.Format(
+                            CultureInfo.CurrentCulture,
+                            Strings.Log_VersionConflict,
+                            versionConflict.Selected.Key.Name)
+                        + $" {Environment.NewLine} {versionConflict.Selected.GetPath()} {Environment.NewLine} {versionConflict.Conflicting.GetPath()}.");
+                    return false;
+                }
+
+                foreach (var downgrade in g.AnalyzeResult.Downgrades)
+                {
+                    var downgraded = downgrade.DowngradedFrom;
+                    var downgradedBy = downgrade.DowngradedTo;
+
+                    // Not all dependencies have a min version, if one does not exist use 0.0.0
+                    var fromVersion = downgraded.Key.VersionRange.MinVersion ?? new NuGetVersion(0, 0, 0);
+                    var toVersion = downgradedBy.Key.VersionRange.MinVersion ?? new NuGetVersion(0, 0, 0);
+
+                    logger.LogWarning(
+                        string.Format(
+                            CultureInfo.CurrentCulture,
+                            Strings.Log_DowngradeWarning,
+                            downgraded.Key.Name,
+                            fromVersion,
+                            toVersion)
+                        + $" {Environment.NewLine} {downgraded.GetPath()} {Environment.NewLine} {downgradedBy.GetPath()}");
+                }
+            }
+
+            return true;
+        }
+
+        private async Task<IEnumerable<ToolRestoreResult>> ExecuteToolRestoresAsync(
+            NuGetv3LocalRepository localRepository,
+            CancellationToken token)
+        {
+            var toolPathResolver = new ToolPathResolver(_request.PackagesDirectory);
+            var results = new List<ToolRestoreResult>();
+
+            foreach (var tool in _request.Project.Tools)
+            {
+                _logger.LogMinimal(string.Format(
+                    CultureInfo.CurrentCulture,
+                    Strings.Log_RestoringToolPackages,
+                    tool.LibraryRange.Name,
+                    _request.Project.FilePath));
+
+                // Build a package spec in memory to execute the tool restore as if it were
+                // its own project. For now, we always restore for a null runtime and a single
+                // constant framework.
+                var toolPackageSpec = new PackageSpec(new JObject())
+                {
+                    Name = string.Empty,
+                    Dependencies = new List<LibraryDependency>(),
```

Comment:
Add the tool as a dependency here so that the lock file will contain the original version range for debugging.